### ISSUE TITLE
no implementation available for O_keywords

### ISSUE THREAD

**Issue by olafhering** [camlp5/camlp5#107]
After upgrading from f776d83014526f654b717e43ed408ec60702aa6b to 080b168b4461b0749f9b96bf9b1ca86d09776d6f,  garrigue/lablgtk@8a803575ded8107a063a8779c47ca79310de90bc fails for me:

````
Running[95]: (cd _build/default/tools && /usr/bin/bash -e -u -o pipefail -c 'camlp5o pr_o.cmo -impl varcc.ml4 -o varcc.ml')
....
File "tools/dune", lines 7-10, characters 0-120:
 7 | (rule
 8 |  (targets varcc.ml)
 9 |  (deps (:source varcc.ml4))
10 |  (action (bash "camlp5o pr_o.cmo -impl %{source} -o %{targets}")))
Command [95] exited with code 2:
$ (cd _build/default/tools && /usr/bin/bash -e -u -o pipefail -c 'camlp5o pr_o.cmo -impl varcc.ml4 -o varcc.ml')
Error while loading "/usr/lib64/ocaml/camlp5/pr_o.cmo": no implementation available for O_keywords.
````

lablgtk has generated files, such as `varcc.ml`, in SCM. The goal is to generate everything from source, my copy of lablgtk does not package such generated files, they are supposed to be generated every time at compile time.

The question is: is this an issue in campl5, or does lablgtk need adjustments for new camlp5?

**Comment by chetmurthy** [comment]
Yes, this a new behaviour of camlp5.  I attach text from camlp5/CHANGES.  Let me know if you need clarification or help implementing this.  [BTW, "not-ocamlfind" is from the package "not-ocamlfind"]

Camlp5 Version 8.03.00:
--------------------

* [25 Apr 2024] update for Ocaml 5.2.0

  Only really salient change (other than supporting new ocaml syntax)
  is that the former Camlp5 revised-syntax "raw-identifier" syntax is
  now removed in favor of the new Ocaml 5.2.0 raw-identifier syntax.
  This should affect almost literally nobody, but just in case, the
  old syntax was (e.g.) "\begin" and the new 5.2.0 syntax is "\#begin".

  I highly doubt that anybody used the old syntax, since it was
  undocumented and only available in revised mode, but hey, just in
  case, letting you know I removed it.

  BUT BUT BUT NOTE WELL this has impact on packages that use camlp5
  internals, b/c some of the internal interfaces have changed.  Specifically:

  (1) if you use "camlp5 pa_r.cmo .... pr_o.cmo" to invoke the preprocessor
  rather than

  "not-ocamlfind preprocess -package camlp5.pa_r,....,camlp5.pr_o"

  you will need to add "o_keywords.cmo" before "pr_o.cmo" (this is
  required by the need to print keywords for each syntax as escaped
  raw identifiers.  Since each syntax has a slightly different set of
  keywords, the printer needs to be parameterized by the
  keyword-table.

  (2) Similarly, if you use the Camlp5 lexer in your package, there is
  a new field, "kwds", which is initialized to the keyword hashtable,
  viz.

    let kwd_table = Hashtbl.create 301 in
    let glex =
    {Plexing.tok_func = func ctx kwd_table glexr;
     tok_using = using_token ctx kwd_table;
     tok_removing = removing_token kwd_table;
     tok_match = tok_match; tok_text = text; tok_comm = None ; kwds = kwd_table }

  note the "kwds = kwd_table".

**Comment by chetmurthy** [comment]
Also, the method I describe works (and has worked) for many releases of camlp5: I forget how long ago, but it has worked since the 8.xx releases I think.  I wrote "not-ocamlfind" specifically to support this usage.

OOF!  Also, I see that I left out that you need to use the same arguments you would with a "ocamlfind ocamlc ...." invocation, including the "-syntax camlp5o" or "-syntax camlp5r"/"-syntax camlp5scm"/etc.  I'll update the CHANGES file.

**Comment by olafhering** [comment]
Thanks, I failed to look into the CHANGES file, sorry for that.

In the meantime I use this construct to inject the new syntax:

````
read < <(camlp5 -version)
case "${REPLY}" in
7*) ;;
8.00.*) ;;
8.02.*) ;;
*)
sed -i~ 's@camlp5o pr_o.cmo -impl@camlp5o o_keywords.cmo pr_o.cmo -impl@' tools/dune
diff -u "$_"~ "$_" && exit 1
;;
esac
````

**Comment by bentxt** [comment]
Thanks alot for the clarification. I went a little nuts because one of my dear commands where failing

```
camlp5 pa_sml.cmo pr_o.cmo -impl foo.sml
Error while loading "~/.opam/5.1.1/lib/camlp5/pr_o.cmo": no implementation available for O_keywords.
```

Now it works again with:
```
 camlp5 pa_sml.cmo o_keywords.cmo  pr_o.cmo -impl foo.sml
```

**Comment by chetmurthy** [comment]
your fixup will work, but please do consider using the `not-ocamlfind preprocess ....` suggestion in the CHANGES file entry.  It has the advantage that it will continue to work in the future regardless of how object files get reorganized.
